src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Serialization.cs

```diff
@@ -57,60 +57,87 @@ internal partial class SymbolTreeInfo : IObjectWritable
             Func<ObjectReader, T> tryReadObject,
             CancellationToken cancellationToken) where T : class, IObjectWritable, IChecksummedObject
         {
-            if (checksum == null) 
+            Func<Task<T>> wrappedTask = async () =>
             {
-                return loadOnly ? null : await createAsync().ConfigureAwait(false);
-            }
+                using (Logger.LogBlock(FunctionId.SymbolTreeInfo_Create, cancellationToken))
+                {
+                    return await createAsync().ConfigureAwait(false);
+                }
+            };
 
-            // Ok, we can use persistence.  First try to load from the persistence service.
-            var persistentStorageService = (IPersistentStorageService2)solution.Workspace.Services.GetService<IPersistentStorageService>();
+            return TryLoadOrCreateWorkerAsync(solution, checksum, loadOnly, wrappedTask, keySuffix, tryReadObject, cancellationToken);
+        }
 
-            T result;
-            using (var storage = persistentStorageService.GetStorage(solution, checkBranchId: false))
+        /// <summary>
+        /// Generalized function for loading/creating/persisting data.  Used as the common core
+        /// code for serialization of SymbolTreeInfos and SpellCheckers.
+        /// </summary>
+        private static async Task<T> TryLoadOrCreateWorkerAsync<T>(
+            Solution solution,
+            Checksum checksum,
+            bool loadOnly,
+            Func<Task<T>> createAsync,
+            string keySuffix,
+            Func<ObjectReader, T> tryReadObject,
+            CancellationToken cancellationToken) where T : class, IObjectWritable, IChecksummedObject
+        {
+            using (Logger.LogBlock(FunctionId.SymbolTreeInfo_TryLoadOrCreate, cancellationToken))
             {
-                // Get the unique key to identify our data.
-                var key = PrefixMetadataSymbolTreeInfo + keySuffix;
-                using (var stream = await storage.ReadStreamAsync(key, cancellationToken).ConfigureAwait(false))
-                using (var reader = ObjectReader.TryGetReader(stream))
+                if (checksum == null)
```

Comment:
I wasn't needed that logging when I was trying to find out how often new indices are created. 

it is always better adding those logging when someone actually does investigation since most of times, what people think what would be needed and what is actually needed for real investigation is often not quite same.

Reply:
I've found the reverse to be true.  When adding logging i do an exercise of simulating getting results back from logging and then trying to figure out what can be done about that.  I've found that doing a little up front analysis can lead to providing better logging data that can be used to make more informed discussions. 

Or, in other words, say i had made this PR but had picked an entirely different set of statements to log.  Why would those statements necessarily be the ones that would lead to valuable changes in the future?  One thing that the devdiv data collection/analysis courses covered was the importance of collecting the right data in order to make appropriate conclusions.  So i'm tyring to apply some of the learnings there to this task :)

Reply:
if one adds logging that was necessary to find out a certain aspect of a system which one actually used to find out the information one needed, then next time, one can use same logging to find out same information if it is needed next.

if you did another investigation for whatever info you need and added bunch of logging that was required to find out info you wanted, that's fine as well since that can be used next time again to find out same info if needed.

if you believe, the same investigation never needed to be done again, then no point of adding the logging.

I left the logging since I believe I will do same investigation again once we settle down on partial load mode thing we are talking about.

Reply:
> I wasn't needed that logging when I was trying to find out how often new indices are created.

This is why it would be important to break things down.  because there are several types of indices, and aggregating all the data together will skew the results.

For example, source-symbol indices are created when the checksum for a project changes.  This checksum is computed using the checksums of all the document-states in the project, combined with the checksums for the compilation and parse options.

However, metadata-symbol indices are updated whenever a PE-reference's properties or mvids change.  

Thees are all triggered on demand if a feature needs them, or automatically by a background incremental analyzer so that they are kept up to date.

--

Note: the spell checker code happens automatically here as well when making a source/metadata index.  But it's unclear to me if you want the spell checker task costs to be included with the other indices.  It will also be difficult to determine the breakdown on things.  For example, effectively all projects will be at a minimum double counted since we make effectively make one logical index, but use two actual calls (which you've added logging to).  It's unclear to me if that was your intent.  It seems like that could confuse numbers and misplace costs.

Reply:
like I said, I wasn't needed that while investigating what I needed.

I started adding logging for all over the places, symbol tree and syntax tree incremental analyzer, metadata, source, precalculated, on demand, checksum creating, cache hit and miss and etc. and removed ones I dont actually using for investigation. and things left are ones that I actually used.

I removed all logging that I end up not actually using.
...

if one believes such logging is important for what he thinks he is going to investigate, add them, I am not going to argue whether that is valuable to him or not. pretty sure one is adding those since it was valuable for what one is doing.

all logging is off by default and one can turn on ones he wants in tools->options->roslyn->performance in roslyndev hive so, it is not like logging one added will force others to use it.  one can add as much logging point as he wants as long as it is actually used for something and help future investigation of same kind.